language/move-vm/runtime/src/runtime.rs

```diff
@@ -24,15 +25,37 @@ use vm::{
     normalized, CompiledModule, IndexKind,
 };
 
+use std::ops::Deref;
+
+// TODO: this is a utility enum and should not be here...
+// Maybe a better place is common/infallible/rwlock.rs
+enum RwLockGuard<'a, T: ?Sized + 'a> {
+    Reader(RwLockReadGuard<'a, T>),
+    Writer(RwLockWriteGuard<'a, T>),
+}
+
+impl<T: ?Sized> Deref for RwLockGuard<'_, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        match self {
+            RwLockGuard::Reader(r) => r.deref(),
+            RwLockGuard::Writer(w) => w.deref(),
+        }
+    }
+}
+
 /// An instantiation of the MoveVM.
 pub(crate) struct VMRuntime {
     loader: Loader,
+    loader_expiry_lock: RwLock<bool>,
```

Comment:
Why is RWlock used here? Is the intention that we want the VMRuntime to be able to be shared across threads? If so the execution result will be non-deterministic right as they really depends on the execution order: If another thread get that lock first, the execution result would be modified. To fix this, we probably neither need a better synchronization method or simply turn the entire VMRuntime to be single threaded by using `RefCell`

Reply:
There are several questions here.

1. Is the intention that we want the VMRuntime to be able to be shared across threads?
- I think that is at least the expectation for the loader, as everything in the loader is already `Mutex` and `Arc` protected. I can't think of a reason why we need mutex-es there if the loader is not intended to be shared across threads.

2. Why RWLock?
> If another thread get that lock first, the execution result would be modified":

I think this is exactly the reason why we need this RWLock here.

With the lock:
- Concurrent module update is not allowed (`writer_lock` is exclusive).
- It is not possible to update a module when someone is using the code cache, vice versa, it is not possible to access the code cache when someone is updating a module (`reader_lock` and `writer_lock` exclusiveness).
- It is not possible to do any transaction after a module is updated. (Well, there is an exception, `execute_function`, see code comments for the details).

We cannot prevent any of the above without this lock. As a result, you will see an even messier result, as I show in the new E2E test cases in the `loader_cache.rs`.

Reply:
Sry lemme further articulate my question. Let's say the VMRuntime gets shared by two threads. Each thread is doing a `publish_module`. Which module gets updated really depends on which thread holds the write lock first. However, this will immediately become a question for potential future parallel exec as the semantics really should be the earlier transaction in the list should get the write lock and the later one should get a Retry. With this API, we might get the Retry on the earilier transaction if threads aren't scheduled carefully.

Reply:
> Each thread is doing a `publish_module`. Which module gets updated really depends on which thread holds the write lock first.

Prior to this PR, Which module gets published really depends on which thread holds the `Mutex` in the loader first.
```
pub(crate) struct Loader {
    scripts: Mutex<ScriptCache>,
    module_cache: Mutex<ModuleCache>, >>> this mutex
    type_cache: Mutex<TypeCache>,
}
```

Reply:
I totally share your concern on the complication to the upcoming parallel exec. And I am fully onboard with the idea that the loader cache should be maintained at a higher level than then MoveVM. In this way, we can implement the MoveVM in a fully sequential manner and leave the cache coherence and parallel exec decision fully to the Diem adapter (or even higher levels).

Reply:
I would say let's not worry too much about parallel execution in the VM right now. It's a problem that hasn't been well defined in the past and one we will look into soon. If the RwLock is required to make the compiler happy then so be it -- as long as we know it's more of a temporary hack.